doc/man/man3/IncronCfg.3
.TH "IncronCfg" 3 "Sat Apr 7 2012" "Version 0.5.10" "incron" \" -*- nroff -*-
.ad l
.nh
.SH NAME
IncronCfg \- 
.PP
Configuration class'\&.  

.SH SYNOPSIS
.br
.PP
.PP
\fC#include <incroncfg\&.h>\fP
.SS "Static Public Member Functions"

.in +1c
.ti -1c
.RI "static void \fBInit\fP ()"
.br
.RI "\fIInitializes default configuration values'\&. \fP"
.ti -1c
.RI "static void \fBLoad\fP (const std::string &rPath)"
.br
.RI "\fILoads configuration values'\&. \fP"
.ti -1c
.RI "static bool \fBGetValue\fP (const std::string &rKey, std::string &rVal)"
.br
.RI "\fIRetrieves a configuration value'\&. \fP"
.ti -1c
.RI "static bool \fBGetValue\fP (const std::string &rKey, int &rVal)"
.br
.RI "\fIRetrieves a configuration value'\&. \fP"
.ti -1c
.RI "static bool \fBGetValue\fP (const std::string &rKey, unsigned &rVal)"
.br
.RI "\fIRetrieves a configuration value'\&. \fP"
.ti -1c
.RI "static bool \fBGetValue\fP (const std::string &rKey, bool &rVal)"
.br
.RI "\fIRetrieves a configuration value'\&. \fP"
.ti -1c
.RI "static std::string \fBBuildPath\fP (const std::string &rPath, const std::string &rName)"
.br
.RI "\fIBuilds a file path'\&. \fP"
.in -1c
.SS "Static Protected Member Functions"

.in +1c
.ti -1c
.RI "static bool \fBParseLine\fP (const char *s, std::string &rKey, std::string &rVal)"
.br
.RI "\fIParses a line a attempts to get a key and a value'\&. \fP"
.ti -1c
.RI "static bool \fBIsComment\fP (const char *s)"
.br
.RI "\fIChecks whether a line is a comment'\&. \fP"
.in -1c
.SH "Detailed Description"
.PP 
Configuration class'\&. 

This class provides access to values loaded from a configuration file (either a explicitly specified one or the default one)'\&. 
.SH "Member Function Documentation"
.PP 
.SS "std::string IncronCfg::BuildPath (const std::string &rPath, const std::string &rName)\fC [static]\fP"
.PP
Builds a file path'\&. This function composes a path from a base path and a file name'\&.
.PP
\fBParameters:\fP
.RS 4
\fIrPath\fP base path 
.br
\fIrName\fP file name 
.RE
.PP
\fBReturns:\fP
.RS 4
full path 
.RE
.PP

.SS "bool IncronCfg::GetValue (const std::string &rKey, std::string &rVal)\fC [static]\fP"
.PP
Retrieves a configuration value'\&. This method attempts to find the appropriate configuration value for the given key and stores it into the given variable'\&.
.PP
\fBParameters:\fP
.RS 4
\fIrKey\fP value key 
.br
\fIrVal\fP retrieved value 
.RE
.PP
\fBReturns:\fP
.RS 4
true = success, false = failure (invalid key) 
.RE
.PP

.SS "bool IncronCfg::GetValue (const std::string &rKey, int &rVal)\fC [static]\fP"
.PP
Retrieves a configuration value'\&. This method attempts to find the appropriate configuration value for the given key and stores it into the given variable'\&.
.PP
\fBParameters:\fP
.RS 4
\fIrKey\fP value key 
.br
\fIrVal\fP retrieved value 
.RE
.PP
\fBReturns:\fP
.RS 4
true = success, false = failure (invalid key) 
.RE
.PP

.SS "bool IncronCfg::GetValue (const std::string &rKey, unsigned &rVal)\fC [static]\fP"
.PP
Retrieves a configuration value'\&. This method attempts to find the appropriate configuration value for the given key and stores it into the given variable'\&.
.PP
\fBParameters:\fP
.RS 4
\fIrKey\fP value key 
.br
\fIrVal\fP retrieved value 
.RE
.PP
\fBReturns:\fP
.RS 4
true = success, false = failure (invalid key) 
.RE
.PP

.SS "bool IncronCfg::GetValue (const std::string &rKey, bool &rVal)\fC [static]\fP"
.PP
Retrieves a configuration value'\&. This method attempts to find the appropriate configuration value for the given key and stores it into the given variable'\&.
.PP
\fBParameters:\fP
.RS 4
\fIrKey\fP value key 
.br
\fIrVal\fP retrieved value 
.RE
.PP
\fBReturns:\fP
.RS 4
true = success, false = failure (invalid key) 
.RE
.PP

.SS "void IncronCfg::Init ()\fC [static]\fP"
.PP
Initializes default configuration values'\&. 
.SS "bool IncronCfg::IsComment (const char *s)\fC [static, protected]\fP"
.PP
Checks whether a line is a comment'\&. \fBParameters:\fP
.RS 4
\fIs\fP text line 
.RE
.PP
\fBReturns:\fP
.RS 4
true = comment, false = otherwise 
.RE
.PP

.SS "void IncronCfg::Load (const std::string &rPath)\fC [static]\fP"
.PP
Loads configuration values'\&. This method attempts to load configuration values from the specified file'\&. If it fails (e'\&.g'\&. the file doesn't exist) the default file is read'\&. As the last resort (for the case the default file can't be loaded) the hard-wired values are used'\&.
.PP
\fBParameters:\fP
.RS 4
\fIrPath\fP configuration file path 
.RE
.PP

.SS "bool IncronCfg::ParseLine (const char *s, std::string &rKey, std::string &rVal)\fC [static, protected]\fP"
.PP
Parses a line a attempts to get a key and a value'\&. \fBParameters:\fP
.RS 4
\fIs\fP text line 
.br
\fIrKey\fP key 
.br
\fIrVal\fP value 
.RE
.PP
\fBReturns:\fP
.RS 4
true = success, false = failure 
.RE
.PP


.SH "Author"
.PP 
Generated automatically by Doxygen for incron from the source code'\&.


CHANGELOG
0.5.10   2012-04-05
* some includes fixed to build correctly with GCC 4.6 (and 4.4-4.5 - #0000385, #0000447)
* some type issues fixed (#0000447)
* EAGAIN handling in poll() (#0000399, #0000686, #0000691)
* Makefile improved as of #0000512
* typo in various files fixed (#0000412)


0.5.9    2009-06-21
* spaces in event-related file names now handled correctly (#0000174)
* crashing on 'incrontab --user' fixed (#0000187)
* man pages improved (bug reporting info as of #0000407)
* Makefile improved ('release-clean' also removes 'doc')


0.5.8    2008-12-29
* more includes added for GCC 4.3 compatibility (#0000178)
* typo in help messages fixed (#0000172)
* some warnings fixed (#0000372)


0.5.7    2007-09-04
* man pages updated (mainly due to #0000157)
* some includes added for GCC 4.3 compatibility (#0000158)
* supplementary groups initialization (#0000167)


0.5.6    2007-04-18
* based on inotify-cxx 0.7.2 (fixes compilation problems on Debian Sarge, #0000155)


0.5.5    2007-02-14
* another bug related to environment variables fixed (#0000151)
* -V (--version) malfunction fixed (#0000150)


0.5.4    2007-02-11
* bug related to persisting environment fixed (#0000149)
* incrond(8) man page updated to explain how the environment is treated
* buggy example configuration file fixed


0.5.3    2007-02-08
* bug related to ignoring new tables fixed (#0000142)
* explicit typecasting in execlp() call (#0000145)
* memory leak related to loaded tables fixed (#0000144)
* documentation for 'incrontab -d' added (#0000146)
* man page for incron.conf added (#0000147)


0.5.2    2007-02-04
* lockfile permissions changed to 0644 (reopened #0000137)
* incorrect reporting of removing non-existing tables fixed (#0000141)


0.5.1    2007-01-28
* lockfile location problems fixed (#0000137)
* lockfile configuration failure fixed (#0000140)
* CRLFs inside LICENCE-GPL changed to LFs (#0000138)


0.5.0    2007-01-27
* based on inotify-cxx 0.7.1
* bug related to events names fixed (#0000134)
* instance locking (only one instance allowed - also fixes #0000136)
* configuration introduced (including custom cfg. files)
* waiting for child processes rewritten to avoid race conditions
* user can request reloading the table (#0000130)
* new implementation for command line arguments
* some code refactoring


0.4.0    2007-01-13
* based on inotify-cxx 0.7.0
* incrontab has a feature to find out supported event types (--types or -t)
* incrond now supports command line arguments
* incrond can be run on foreground (--foreground, -n)
* incrond can be simply stopped (--kill, -k)
* support for system tables (/etc/incron.d) added
* eaccess() replaced by euiaccess() (#0000125)
* each table now use its own inotify descriptor (#0000129)
* descriptors are closed on executing child processes


0.3.5    2007-01-09
* based on inotify-cxx 0.6.3
* a security bug related to access rights to watched files fixed (#0000119)
* crashing on table reload after removing a watched path fixed (#0000123)
* 'incrontab --list' dysfunction fixed (#0000122)


0.3.4    2007-01-03
* based on inotify-cxx 0.6.2
* problems with wrong ordered setting of UID/GID fixed (#0000115, #0000117)
* user tables have 0600 permission now


0.3.3    2006-12-15
* based on inotify-cxx 0.5.3
* "GID ignorance" bug fixed (#0000109)
* "zombie bug" fixed (#0000105)
* code cleanup and simplifying


0.3.2    2006-12-02
* problems with reiserfs fixed (#0000104)


0.3.1    2006-11-24
* user manual added
* incron tables now owned by root (also fixes #0000103)


0.3.0    2006-11-12
* based on inotify-cxx 0.5.2 (fixes a problem with ignoring IN_OPEN)
* two new special symbols ($% for flags as names, $& for numeric flags)
* whitespaces in paths now allowed (by prepending a backslash) (#0000098)
* some changes in logging etc.


0.2.3    2006-10-30
* problems with unwanted IN_NO_LOOP have been fixed (#0000097)


0.2.2    2006-10-29
* based on inotify-cxx 0.5.0
* 'uncleaned children' bug has been fixed (#0000095)
* memory leak (child process arguments) has been fixed (#0000096)


0.2.1    2006-10-14
* based on inotify-cxx 0.4.1 (it should also fix crashing as described in #0000085)
* build failure (due to missing 'useradd -M') on some distributions has been fixed
  (#0000094)
* tarball structure has been fixed (#0000089)
* documentation slightly improved


0.2.0    2006-10-04
* based on inotify-cxx 0.3.1
* better signal handling (no "dirty" wait() in a handler)
* loop avoidance can be used
* more documentation in the code


0.1.0    2006-09-15
* first alpha version


incroncfg.cpp

/// inotify cron configuration implementation
/**
 * \file incroncfg.cpp
 * 
 * incron configuration
 * 
 * Copyright (C) 2007, 2008, 2012 Lukas Jelinek, <lukas@aiken.cz>
 * 
 * This program is free software; you can use it, redistribute
 * it and/or modify it under the terms of the GNU General Public
 * License, version 2 (see LICENSE-GPL).
 *  
 * Credits:
 *   Christian Ruppert (new include to build with GCC 4.4+)
 * 
 */


#include <fstream>
#include <sstream>
#include <cstring>

#include "incroncfg.h"


#define INCRON_CFG_DEFAULT "/etc/incron.conf"


typedef std::map<std::string, std::string> CFG_MAP;
typedef CFG_MAP::iterator CFG_ITER;


CFG_MAP IncronCfg::m_values;
CFG_MAP IncronCfg::m_defaults;


void IncronCfg::Init()
{
  m_defaults.insert(CFG_MAP::value_type("system_table_dir", "/etc/incron.d"));
  m_defaults.insert(CFG_MAP::value_type("user_table_dir", "/var/spool/incron"));
  m_defaults.insert(CFG_MAP::value_type("allowed_users", "/etc/incron.allow"));
  m_defaults.insert(CFG_MAP::value_type("denied_users", "/etc/incron.deny"));
  m_defaults.insert(CFG_MAP::value_type("lockfile_dir", "/var/run"));
  m_defaults.insert(CFG_MAP::value_type("lockfile_name", "incrond"));
  m_defaults.insert(CFG_MAP::value_type("editor", ""));
}

void IncronCfg::Load(const std::string& rPath)
{
  char s[1024];
  
  std::ifstream is(rPath.c_str());
  if (is.is_open()) {
    while (!is.eof() && !is.fail()) {
      is.getline(s, 1023);
      std::string key, val;
      if (ParseLine(s, key, val)) {
        m_values.insert(CFG_MAP::value_type(key, val));
      }
    }
    is.close();
    return;
  }
  
  if (rPath == INCRON_CFG_DEFAULT)
    return;
  
  is.open(INCRON_CFG_DEFAULT);
  if (is.is_open()) {
    while (!is.eof() && !is.fail()) {
      is.getline(s, 1023);
      std::string key, val;
      if (ParseLine(s, key, val)) {
        m_values.insert(CFG_MAP::value_type(key, val));
      }
    }
    is.close();
  }
}

bool IncronCfg::GetValue(const std::string& rKey, std::string& rVal)
{
  CFG_ITER it = m_values.find(rKey);
  if (it != m_values.end()) {
    rVal = (*it).second;
    return true;  
  }
  
  it = m_defaults.find(rKey);
  if (it != m_defaults.end()) {
    rVal = (*it).second;
    return true;  
  }
  
  return false;
}

bool IncronCfg::GetValue(const std::string& rKey, int& rVal)
{
  std::string s;
  if (GetValue(rKey, s)) {
    if (sscanf(s.c_str(), "%i", &rVal) == 1)
      return true; 
  }
  
  return false;
}

bool IncronCfg::GetValue(const std::string& rKey, unsigned& rVal)
{
  std::string s;
  if (GetValue(rKey, s)) {
    if (sscanf(s.c_str(), "%u", &rVal) == 1)
      return true; 
  }
  
  return false;
}

bool IncronCfg::GetValue(const std::string& rKey, bool& rVal)
{
  std::string s;
  if (GetValue(rKey, s)) {
    size_t len = (size_t) s.length();
    for (size_t i = 0; i < len; i++) {
      s[i] = (char) tolower(s[i]);
    }
    
    rVal = (s == "1" || s == "true" || s == "yes" || s == "on" || s == "enable" || s == "enabled");
    return true;
  }
  
  return false;
}

std::string IncronCfg::BuildPath(const std::string& rPath, const std::string& rName)
{
  if (rPath.rfind('/') == rPath.length() - 1)
    return rPath + rName;
    
  return rPath + "/" + rName;
}

bool IncronCfg::ParseLine(const char* s, std::string& rKey, std::string& rVal)
{
  // CAUTION: This code hasn't been optimized. It may be slow.
  
  char key[1024], val[1024];
  
  if (IsComment(s))
    return false;
  
  std::istringstream ss(s);
  ss.get(key, 1023, '=');
  if (ss.fail())
    return false;
    
  ss.get(val, 1023);
  if (ss.fail())
    return false;
    
  rKey = key;
  rVal = val;
    
  std::string::size_type a = rKey.find_first_not_of(" \t");
  std::string::size_type b = rKey.find_last_not_of(" \t");
  if (a == std::string::npos || b == std::string::npos)
    return false;
    
  rKey = rKey.substr(a, b-a+1);
  
  a = rVal.find_first_not_of(" \t=");
  b = rVal.find_last_not_of(" \t");
  if (a == std::string::npos || b == std::string::npos) {
    rVal = "";
  }
  else {
    rVal = rVal.substr(a, b-a+1);
  }
    
  return true;
}

bool IncronCfg::IsComment(const char* s)
{
  const char* sx = strchr(s, '#'); // 
  if (sx == NULL)
    return false;
    
  size_t len = sx - s;
  for (size_t i = 0; i < len; i++) {
    if (!(s[i] == ' ' || s[i] == '\t'))
      return false;
  }
  
  return true;
}

